pyphi/tpm.py

```diff
@@ -50,11 +50,35 @@ def condition_tpm_nb(
     tpm, fixed_nodes, state, num_states_per_node=None, node_labels=None
 ):
     df = tpm2df(tpm, num_states_per_node, node_labels)
-    for c in fixed_nodes:
-        df = df.iloc[df.index.get_level_values(c) == state[node_labels.index(c)]]
-    free_nodes = list(set(node_labels) - set(fixed_nodes))
+    num_states_per_node = list(num_states_per_node)
+
+    for c in reversed(fixed_nodes):
+        bool_array = [value == state[c] for value in df.index.get_level_values(c)]
+        df = df.iloc[bool_array]
+        del num_states_per_node[c]
```

Comment:
What you've done is a good adaptation of the original code, but there is a chance to improve this significantly by using more idiomatic pandas. I would try to do the conditioning in a single step by constructing a tuple indexer, and then making one call to `.loc`.

See [the indexing docs](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html#advanced-indexing-with-hierarchical-index). The [`pd.IndexSlice`]( https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.IndexSlice.html#pandas.IndexSlice) object would be useful here.